tests/end_to_end/end_to_end.cpp

```diff
@@ -673,3 +680,106 @@ TEST_CASE("pinned_map_enum", "[end_to_end]")
+         error_message = nullptr,
+         result == EBPF_SUCCESS));
+
+    REQUIRE(hook.attach(program_handle) == EBPF_SUCCESS);
+
+    // Close program handle. That should detach the program from the hook
+    // and unload the program.
+    ebpf_api_close_handle(program_handle);
+    program_handle = INVALID_HANDLE_VALUE;
+    REQUIRE(ebpf_api_get_next_program(program_handle, &program_handle) == EBPF_SUCCESS);
+    REQUIRE(program_handle == INVALID_HANDLE_VALUE);
+
+    // Close link handle. This should delete the link object.
+    hook.close_handle();
+
+    link_handle = INVALID_HANDLE_VALUE;
+    REQUIRE(ebpf_core_get_next_handle(link_handle, EBPF_OBJECT_LINK, &link_handle) == EBPF_SUCCESS);
+    REQUIRE(link_handle == INVALID_HANDLE_VALUE);
+
+    REQUIRE(
+        (result = ebpf_api_load_program(
+             SAMPLE_PATH "droppacket.o",
+             "xdp",
+             EBPF_EXECUTION_INTERPRET,
+             &program_handle,
+             &count_of_map_handle,
+             &map_handle,
+             &error_message),
+         error_message ? printf("ebpf_api_load_program failed with %s\n", error_message) : 0,
+         ebpf_free_string(error_message),
+         error_message = nullptr,
+         result == EBPF_SUCCESS));
+
+    REQUIRE(hook.attach(program_handle) == EBPF_SUCCESS);
+    // Close program handle. That should detach the program from the hook
+    // and unload the program.
+    ebpf_api_close_handle(program_handle);
+    program_handle = INVALID_HANDLE_VALUE;
+    REQUIRE(ebpf_api_get_next_program(program_handle, &program_handle) == EBPF_SUCCESS);
+    REQUIRE(program_handle == INVALID_HANDLE_VALUE);
+
+    // Detach and close link handle.
+    hook.detach();
+
+    link_handle = INVALID_HANDLE_VALUE;
+    REQUIRE(ebpf_core_get_next_handle(link_handle, EBPF_OBJECT_LINK, &link_handle) == EBPF_SUCCESS);
+    REQUIRE(link_handle == INVALID_HANDLE_VALUE);
+
+    REQUIRE(
+        (result = ebpf_api_load_program(
+             SAMPLE_PATH "droppacket.o",
+             "xdp",
+             EBPF_EXECUTION_INTERPRET,
+             &program_handle,
+             &count_of_map_handle,
+             &map_handle,
+             &error_message),
+         error_message ? printf("ebpf_api_load_program failed with %s\n", error_message) : 0,
+         ebpf_free_string(error_message),
+         error_message = nullptr,
+         result == EBPF_SUCCESS));
+
+    REQUIRE(hook.attach(program_handle) == EBPF_SUCCESS);
+
+    // Detach and close link handle.
+    hook.detach();
+
+    link_handle = INVALID_HANDLE_VALUE;
+    REQUIRE(ebpf_core_get_next_handle(link_handle, EBPF_OBJECT_LINK, &link_handle) == EBPF_SUCCESS);
+    REQUIRE(link_handle == INVALID_HANDLE_VALUE);
+
+    // Close program handle.
+    ebpf_api_close_handle(program_handle);
+    program_handle = INVALID_HANDLE_VALUE;
+    REQUIRE(ebpf_api_get_next_program(program_handle, &program_handle) == EBPF_SUCCESS);
+    REQUIRE(program_handle == INVALID_HANDLE_VALUE);
```

Comment:
how do you test the scenario, load -> pin -> attach -> exit -> open -> detach/unload(close) etc. ?

Reply:
That test is not present yet. But I think all the APIs needed for testing the above flow are also not present yet.